http/src/main/scala/org/http4s/blaze/http/http2/StreamManager.scala

```diff
@@ -1,11 +1,82 @@
 package org.http4s.blaze.http.http2
 
-// Place holder
+import org.http4s.blaze.http._
+
+import scala.concurrent.Future
+
+/** Manager of the active streams for a session
+  *
+  * The `StreamManager` can be thought of as the collection of
+  * active streams and some associated helper methods for performing
+  * operations relevant to the HTTP/2 protocol.
+  */
 private trait StreamManager {
 
+  /** The `StreamIdManager` owned by this `StreamManager` */
+  def idManager: StreamIdManager
+
+  /** Number of active streams */
   def size: Int
 
-  def streamFinished(stream: StreamState, cause: Option[Http2Exception]): Unit
+  /** Whether any streams are active */
+  def isEmpty: Boolean
+
+  /** Update the flow windows of open streams due to a change of the initial flow window
+    *
+    * A receiver MUST adjust the size of all stream flow-control windows that
+    * it maintains by the difference between the new value and the old value.
+    * https://tools.ietf.org/html/rfc7540#section-6.9.2
+    *
+    * @param delta difference between the new initial window and the previous initial window
+    */
+  def initialFlowWindowChange(delta: Int): MaybeError
+
+  /** Get the stream associated with the specified stream ID */
+  def get(streamId: Int): Option[StreamState]
+
+  /** Register an `InboundStreamState` with the set of active streams */
+  def registerInboundStream(state: InboundStreamState): Boolean
 
+  /** Register an `OutboundStreamState` with the set of active streams
+    *
+    * @return the newly allocated stream id to be associated with the new
+    *         outbound stream, it it was possible to allocate one.
+    */
   def registerOutboundStream(state: OutboundStreamState): Option[Int]
+
+  /** Cause the associated stream to be reset, if it exists
+    *
+    * @param cause the reason the stream was reset
+    * @return true if the stream existed and was closed, false otherwise
+    */
+  def rstStream(cause: Http2StreamException): Boolean
+
+  /** Called by a `StreamState` to remove itself from the StreamManager
+    *
+    * @param streamState the stream being closed
+    * @return true if the stream existed and was closed, false otherwise
+    */
+  def streamClosed(streamState: StreamState): Boolean
+
+  /** Handle a valid and complete PUSH_PROMISE frame */
+  def handlePushPromise(streamId: Int, promisedId: Int, headers: Headers): Http2Result
+
+  /** Update the specified flow window with the specified bytes
+    *
+    * @note stream ID 0 indicates the session flow window
+    */
+  def flowWindowUpdate(streamId: Int, sizeIncrement: Int): MaybeError
+
+  /** Close the `StreamManager` and all the associated streams immediately */
+  def close(cause: Option[Throwable]): Unit
```

Comment:
I didn't add a `Future[Unit]` return type because nobody cares when its done: this is method for freeing stream related resources during a session panic. I should make the docs more clear about that, and maybe the name as well.

Technically, it could be arbitrarily slow since the close will propagate down to stream pipelines which can then perform arbitrary work and may not be forking their Future's into a different thread. I consider that to be the responsibility of the user to use an `ExecutionContent` as appropriate, and shouldn't be a big deal.